flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java

```diff
@@ -0,0 +1,197 @@
+                context.path = path;
+                return WaitForJobTermination;
+            }
+
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                context.unfinishedStates = extractUnfinishedStates(executionStates);
+                return WaitForSavepointCreation;
+            }
+        },
+        WaitForSavepointCreation {
+            @Override
+            public StopWithSavepointState handleSavepointCreation(
+                    StopWithSavepointContext context, String path) {
+                Preconditions.checkState(
+                        context.unfinishedStates != null,
+                        InitialWait + " should have preceded: No unfinishedStates is set.");
+
+                if (!context.unfinishedStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(context.unfinishedStates);
+                }
+
+                return context.terminateSuccessfully(path);
+            }
+        },
+        WaitForJobTermination {
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                Preconditions.checkState(
+                        context.path != null,
+                        InitialWait + " should have preceded: No path is set.");
+
+                Collection<ExecutionState> unfinishedExecutionStates =
+                        extractUnfinishedStates(executionStates);
+                if (!unfinishedExecutionStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(unfinishedExecutionStates);
+                }
+
+                return context.terminateSuccessfully(context.path);
+            }
+        },
+        Final;
+
+        public StopWithSavepointState onSavepointCreation(
+                StopWithSavepointContext context, String path, Throwable throwable) {
+            if (throwable != null) {
+                return context.terminateExceptionally(throwable);
+            }
+
+            return handleSavepointCreation(context, path);
+        }
+
+        protected StopWithSavepointState handleSavepointCreation(
+                StopWithSavepointContext context, String path) {
+            throw new IllegalStateException(
+                    "No onSavepointCreation should have been called in " + this.name() + " state.");
+        }
+
+        public StopWithSavepointState onExecutionsTermination(
+                StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+            throw new IllegalStateException(
+                    "No onExecutionsTermination should have been called in "
+                            + this.name()
+                            + " state.");
+        }
```

Comment:
I think this does not work because a savepoint can fail and a later point in time `onExecutionsTermination` could be called (e.g. if an unrelated global failover is triggered).

Reply:
You're right. I added all possible test cases to the most recent code version.